### PR TITLE
[Perf fw] add test metadata to the tests for the perf fw

### PR DESCRIPTION
**Pull request by vhvb1989**
fixes: #1636 

Adding test metadata class and letting each test to define it.
This data is used to print the test info when running the test

### CODE REVIEW

File: sdk/core/performance-stress/inc/azure/performance-stress/test_metadata.hpp [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+#include <functional>
+#include <memory>
+#include <string>
+
+namespace Azure { namespace PerformanceStress {
+  /**
+   * @brief Define the metadata of a test that can be run by the performance framework.
+   *
+   */
+  struct TestMetadata
```
Comment (RickWinter) [inline sdk/core/performance-stress/inc/azure/performance-stress/test_metadata.hpp:24]: I like this design for encapsolating the Test.  However I would stick with a hash map of tests.  Most of the uses need to do a lookup and looking up with a unique name is more efficient than iterating over a list doing string comparisons.

To make the code easy to use we can define a type for it
```
using TestCollection = std::map<
        std::string,
        std::function<std::unique_ptr<Azure::PerformanceStress::PerformanceTest>(
            Azure::PerformanceStress::TestOptions)>>;
```
Then everywhere we use it we can do:
```
inline std::unique_ptr<Azure::PerformanceStress::PerformanceTest> PrintAvailableTests(
    TestCollection const& tests)
```

Comment (RickWinter) [reply sdk/core/performance-stress/inc/azure/performance-stress/test_metadata.hpp:24]: You could go a step further and make the collection a class with a private member hash.  Then add helpers over the top to get/set tests into it.

Comment (vhvb1989) [reply sdk/core/performance-stress/inc/azure/performance-stress/test_metadata.hpp:24]: >  Most of the uses need to do a lookup and looking up with a unique name

We only need one search per application-run, and the number of the tests inside the application will never be that big (I would be impressed if it ever get to 100).  The improvement of using a Set for the O(1)  vs/ the O(n) in the Vector won't make a significant difference.

Also, looking for the test is not something impacting running the actual test.

**Comment by vhvb1989** [comment]
closing in favor of only reviewing https://github.com/Azure/azure-sdk-for-cpp/pull/1655 as people have started to look on it

**Pull request closed without merging**